tmt/base.py

```diff
@@ -1529,6 +1529,45 @@ def lint_require_type_field(self) -> LinterReturn:
         yield LinterOutcome.FIXED, 'added type to requirements'
 
 
+@dataclasses.dataclass(repr=False)
+class Tests(tmt.lint.Lintable['Tests']):
+    """ List of Test objects for linting """
+
+    def __init__(self, tests: list['tmt.Test'], *args: Any, **kwargs: Any) -> None:
+        super().__init__(*args, **kwargs)
+        self.tests = tests
+
+    @classmethod
+    def from_tree(cls, tree: 'tmt.Tree') -> 'Tests':
+        return Tests(tree.tests())
+    
+    def lint_no_duplicate_test_ids(self) -> LinterReturn:
+        """ X001: no duplicate test ids """
+
+        duplicate_ids: dict[str, list[str]] = dict()
+        for test in self.tests:
+            if test.id is None:
+                continue
+            if test.id not in duplicate_ids:
+                duplicate_ids[test.id] = []
+            duplicate_ids[test.id].append(test.name)
+
+        has_duplicates = False
+        for id, test_names in duplicate_ids.items():
+            if len(test_names) > 1:
+                has_duplicates = True
+                for name in test_names:
+                    yield LinterOutcome.FAIL, f'duplicate id "{id}" in test "{name}"'
+
+        if has_duplicates:
+            return
```

Comment:
We could simplify a bit, how about the following? (`id` would conflict with a built-in function, hence calling it `test_id`, no need for an extra flag, `defaultdict` to simplify collecting phase:

```python
ids: collection.defaultdict[str, list[str]] = collection.defaultdict(list)

for test in self.tests:
    if test.id is None:
        continue
    ids[test.id].append(test.name)

duplicates = {
    test_id: test_names for test_name, test_ids in duplicate_ids.items() if len(test_names) > 1
}

for test_id, test_names in duplicates.items():
    for name in test_names:
        yield LinterOutcome.FAIL, f'duplicate id "{test_id}" in test "{name}"'

if duplicates:
    return
```

Reply:
Looks great, changed it accordingly.